File: 139-word-break/word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& dict) {
        int n= (int)s.size();
        vector<bool> dp(n+1,false);
        dp[0]=true;
        for(int i=0;i<n;++i){
            if(!dp[i])  continue;
            for(int j=0;j<(int)dict.size();++j){
                int b=0;
                while(i+b<n && b<dict[j].size() && s[i+b]==dict[j][b]){
                    ++b;
                }
                if(b==dict[j].size()){
                    dp[i+b]=true;
                }
            }
        }
        return dp[n];
    }
};

File: 139-word-break/word-break.java
class TrieNode{
    TrieNode[] children;
    boolean isEnd;
    public TrieNode(){
        children= new TrieNode[26];
        isEnd= false;
    }
}
class Solution {
    
    private void insertWord(String word, TrieNode curr){
        for(char ch: word.toCharArray()){
            if(curr.children[ch-'a']==null){
                curr.children[ch-'a']= new TrieNode();
            } 
            curr= curr.children[ch-'a'];
        }
        curr.isEnd= true;
    }
    public boolean wordBreak(String s, List<String> dict) {
        TrieNode root = new TrieNode();
        TrieNode curr;
        for(String word: dict){
            insertWord(word, root);
        }
        int n= s.length();
        boolean[] dp = new boolean[n+1];
        dp[0]=true;
        for(int i=0;i<n;i++){
            if(dp[i]==false) continue;
            int j=i;
            curr= root;
            while(j<n && curr.children[s.charAt(j)-'a']!=null){
                curr= curr.children[s.charAt(j)-'a'];
                if(curr.isEnd){
                    dp[j+1]= true;
                }
                j++;
            }
        }
        return dp[n];
    }
}